text/3177-apply-patchfiles-to-dependencies.md

```diff
@@ -0,0 +1,328 @@
+to upgrade to, since they include way to many breaking changes and additional features, just updating the software to the newest version is 
+not a priority right now. You can now instead just get the specific bugfix included in your software for what you are maintaining. 
+
+``` bash
+git clone <sourcecode>
+git checkout <old-version-hash>
+git cherry-pick <bugfix-hash>
+git diff HEAD~1 > fix-foobarize-bug.patch
+```
+
+And then update your dependency to apply the patch you just created onto that old version of the
+dependency.
+
+``` toml
+# your crate
+[dependencies]
+# apply the created patchfile to the legacy version of foo, that the maintainer do not wish to fix.
+foo = "1.0"
+
+[patch.crates-io]
+foo = { path = "../path/to/foo", patchfiles = ["patches/fix-foobarize-bug.patch"]  }
+```
+
+It could also be used to change the behavior of a dependency in a way that is only useful for
+your application and you know would never be merged into the original software. The patchfiles will be applied
+after `git` or `path` has been resolved if there are any present. Otherwise since patchfiles do not contain any version information, the `version` tag needs
+to be present. 
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+> This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+> - Its interaction with other features is clear.
+> - It is reasonably clear how the feature would be implemented.
+> - Corner cases are dissected by example.
+
+> The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work.
+
+There was initial work created in PR: 
+
+The Cargo `[patch]` section gets information about which source to patch from the Cargo.toml inside the `path` or `git` tags. Since a patchfile can change this information, 
+and do not contain this information in itself, then either the version to patch needs to be gathered from a `path` or `git` tag, but if none of those are supplied, the 
+`version` tag _must_ be specified. The 'version' tag will in this RFC will not interfere with the workings of how versions were previously handled. But is suggested in 
+future work. 
+
+To maintain compatibility with how patches applied previously and to limit the scope of this RFC, the added `version` key in the `[patch]` section will only be allowed 
+and required when no `path` or `git` exists.
+
+The feature would add an extra step to the build process, right after the download of the crate from
+crates.io, path or git. 
+
+It would apply the patch from the patchfiles in the order they are defined in the patches section of
+the dependency declaration, if a patch does not apply cleanly, it would result in an compilation error
+that shows which patch that did not apply cleanly to which library. This would make sure that patchfiles are not to dangerous. 
```

Comment:
```suggestion
the dependency declaration, if a patch has any "fuzz" (mismatched context lines), it would result in an compilation error
that shows which patch that did not apply cleanly to which library. This would make sure that patchfiles are not to dangerous. 
```

Reply:
@joshtriplett imho it should also specify other mismatches too otherwise it'd technically be legal to not fail if a patch deletes lines that don't match since they're not context lines